ado-team-project option on generate-script

DevStarOps opened the pull request:
Adds ado-team-project option on generate-script #124 and updates service connection to be case insensitive as it was not finding the service connection in my org #242 

- [x] Did you write/update appropriate tests
- [x] Release notes updated (if appropriate)
- [x] Appropriate logging output
- [x] Issue linked

dylan-smith commented on src/Octoshift/AdoApi.cs:
```diff
@@ -85,7 +85,7 @@ public virtual async Task<string> GetGithubAppId(string org, string githubOrg, I
                 var response = await _client.GetWithPagingAsync(url);
 
-                var endpoint = response.FirstOrDefault(x => (string)x["type"] == "GitHub" && ((string)x["name"]).ToLower() == githubOrg.ToLower());
+                var endpoint = response.FirstOrDefault(x => ((string)x["type"]).Equals("GitHub", StringComparison.OrdinalIgnoreCase) && ((string)x["name"]).Equals(githubOrg, StringComparison.OrdinalIgnoreCase));
```
Is this related to the ado-team-project change, or is this just a bit of unrelated cleanup?

DevStarOps replied:
I created #242 because of a warning that is thrown at the end of the process where type returned is all lower case for me so was not matching GitHub

dylan-smith commented on src/ado2gh/Commands/GenerateScriptCommand.cs:
```diff
@@ -161,16 +167,18 @@ public async Task Invoke(string githubOrg, string adoOrg, FileInfo output, bool
+                        if (teamProjects.Any(o => o.Equals(adoTeamProject, StringComparison.OrdinalIgnoreCase)))
                         {
-                            _log.LogInformation($"    Repo: {repo}");
+                            await GetTeamProjectRepos(ado, repos, org, adoTeamProject);
                         }
```
```suggestion
                            await GetTeamProjectRepos(ado, repos, org, adoTeamProject);
                        }
                        else
                        {
                            _log.LogWarning($"ADO Team Project provided cannot be found [{adoTeamProject}]");
                        }
```

DevStarOps replied:
I thought about logging a warning but if you have 3 orgs that it's looping over and the team project exists in 1 of them then you'd get 2 warnings thinking something is wrong but it worked as expected

DevStarOps replied:
What I'll do is capture if a team project was found and outside the org loop log the warning, same same, just a little more tracking code :)

dylan-smith commented on src/ado2gh/Commands/GenerateScriptCommand.cs:
```diff
@@ -179,6 +187,18 @@ public async Task Invoke(string githubOrg, string adoOrg, FileInfo output, bool
             return repos;
         }
 
+        private async Task GetTeamProjectRepos(AdoApi ado, Dictionary<string, IDictionary<string, IEnumerable<string>>> repos, string org, string teamProject)
```
I'd suggest returning an IEnumerable<string> and adding it to the repos Dict in the caller instead of passing in repos here.

dylan-smith commented on src/gei/Commands/GenerateScriptCommand.cs:
```diff
@@ -87,7 +93,7 @@ public async Task Invoke(string githubSourceOrg, string adoSourceOrg, string git
 
             var script = string.IsNullOrWhiteSpace(githubSourceOrg) ?
-                await InvokeAdo(adoSourceOrg, githubTargetOrg, ssh) :
+                await InvokeAdo(adoSourceOrg, adoTeamProject, githubTargetOrg, ssh) :
```
We should validate that the user hasn't passed in an Ado Team Project without an Ado Org, and throw an OctoshiftCliException if they did.

DevStarOps replied:
The check above says that if noth github org and ado org are empty it throws.
If both are supplied it goes the GitHub repos root, so if team project is supplied without org, then it's either missing GitHub org too and will throw or will exist

dylan-smith commented:
I'm learning our CI doesn't work too well on PR's from forks. I'll file an issue to fix that later.

dylan-smith commented on src/OctoshiftCLI.Tests/AdoApiTests.cs:
```diff
@@ -238,6 +239,7 @@ public async Task GetGithubAppId_Should_Return_Null_When_No_Team_Projects_Have_E
             var githubOrg = "foo-gh-org";
             var teamProject1 = "foo-tp1";
             var teamProject2 = "foo-tp2";
+            var teamProjectArg = string.Empty;
```
depending on what we do with the previous comment, we may want a test for the scenario where teamProjectArg is not empty.

dylan-smith commented on src/ado2gh/Commands/GenerateScriptCommand.cs:
```diff
@@ -171,36 +177,61 @@ public async Task Invoke(string githubOrg, string adoOrg, FileInfo output, bool
 
+        private async Task<IEnumerable<string>> GetTeamProjectRepos(AdoApi ado, string org, string teamProject)
+        {
+            _log.LogInformation($"  Team Project: {teamProject}");
```
I think this is going to result in redundant log output with line 207

DevStarOps replied:
updated, thanks, was the same on line 197 🙂

dylan-smith commented on src/gei/Commands/GenerateScriptCommand.cs:
```diff
@@ -118,6 +124,7 @@ public GenerateScriptCommand(OctoLogger log, ISourceGithubApiFactory sourceGithu
             if (!string.IsNullOrWhiteSpace(adoSourceOrg))
             {
                 _log.LogInformation($"ADO SOURCE ORG: {adoSourceOrg}");
+                _log.LogInformation($"ADO TEAM PROJECT: {adoTeamProject}");
             }
```
```suggestion
            }
            
            if (!string.IsNullOrWhiteSpace(adoTeamProject))
                _log.LogInformation($"ADO TEAM PROJECT: {adoTeamProject}");
            }
```

dylan-smith commented on src/gei/Commands/GenerateScriptCommand.cs:
```diff
@@ -435,6 +448,18 @@ exit 1
 
+        private async Task<IEnumerable<string>> GetTeamProjectRepos(AdoApi adoApi, string adoOrg, string teamProject)
+        {
+            _log.LogInformation($"Team Project: {teamProject}");
```
duplicate log lines with line 232

DevStarOps replied:
removed from earlier line so that both branches are covered by the method logging

dylan-smith reviewed (changes requested):
The main change I think we need to make is how it handles finding the GH Pipelines App ID when scoped to a team project.  Once we decide that I can make all the other changes and merge it in.

DevStarOps commented:
sorted all the issues, I think a lot of the duplicate logging I messed up during the merge as I added logging where I thought I was missing it forgetting I moved it to the method

For AppId I removed the restriction to only get the service connection for the team project supplied as that's meant to restrict repos and feels miss used and creates a harder migration process
